### PR TITLE
[Tensorpipe Agent] Base Structs for Tracking RPC Metrics

### PR DESCRIPTION
**Pull request by osalpekar**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37980 [Tensorpipe Agent] Implementing getMetrics with currently available metrics
* #37852 [Tensorpipe Agent] Network Data Profiling
* #37851 [Tensorpipe Agent] Implement Global Interpreter Lock Wait Time Metric
* **#37850 [Tensorpipe Agent] Base Structs for Tracking RPC Metrics**

Adding the base structs for tracking time-series metrics in the Tensorpipe RPC Agent

Differential Revision: [D21339520](https://our.internmc.facebook.com/intern/diff/D21339520/)

### CODE REVIEW

File: torch/csrc/distributed/rpc/tensorpipe_agent.h [hunk 1]
```diff
@@ -120,6 +120,25 @@ class TensorPipeAgent : public RpcAgent {
...
+  // Struct for capturing Time-Series Metrics
+  struct TimeSeriesMetricsTracker {
+    uint64_t currentSum_;
+    uint64_t currentCount_;
+
+    explicit TimeSeriesMetricsTracker(
+        uint64_t currentSum = 0,
+        uint64_t currentCount = 0);
+
+    void addData(uint64_t dataPoint);
```
Comment (mrshenli) [inline torch/csrc/distributed/rpc/tensorpipe_agent.h:133]: What is a `dataPoint`?

**Review by mrshenli — COMMENTED** [review]
Are these metrics specific to TensorPipe or do they also applicable to other types of RPC agents?

**Comment by osalpekar** [comment]
> Are these metrics specific to TensorPipe or do they also applicable to other types of RPC agents?

These metrics are applicable to the other agents - most of this metrics logic is adapted from PGA and Thrift (at least for GIL Wait Time and Network Info). Soon, I'll be adding more metrics to acheive parity with the other 2 agents + brainstorm if any other Tensorpipe-specific metrics would be useful.

File: torch/csrc/distributed/rpc/tensorpipe_agent.h (hunk 1 above)
Comment (osalpekar) [reply torch/csrc/distributed/rpc/tensorpipe_agent.h:133]: A datapoint is basically a new observation for the metric we are tracking (variable naming borrowed from equivalent struct in PGA here: https://github.com/pytorch/pytorch/blob/a3639fa5167e1ce88afe2f775c3fa95e4ae646e0/torch/csrc/distributed/rpc/process_group_agent.h#L126).

Comment (jiayisuse) [reply torch/csrc/distributed/rpc/tensorpipe_agent.h:141]: I see we have a TODO in process_group_agent,
  // TODO: this class should inherit from a MetricsTracker, and can be extended
  // to track num_sends, recvs, average size of messages, etc.
Are we going to do the same with tensorpipe agent? :)

Comment (osalpekar) [reply torch/csrc/distributed/rpc/tensorpipe_agent.h:141]: Yeah I noticed that TODO as well - I'm guessing this was to clean up the abstractions for tracking metrics to unify across all agents. Maybe we could keep this in-line with the other agents for now, until this is refactored elsewhere?

Comment (jiayisuse) [reply torch/csrc/distributed/rpc/tensorpipe_agent.h:141]: Yeah we can keep as it is and look at back later

**Review by jiayisuse — APPROVED** [review]

**Review by lw — APPROVED** [review]

File: torch/csrc/distributed/rpc/tensorpipe_agent.h [hunk 2]
```diff
@@ -120,6 +120,34 @@ class TensorPipeAgent : public RpcAgent {
...
+
+    // Adds a data point (which is basically one observation for the metric
+    // being tracked) to the running sum and count.
+    void addData(uint64_t dataPoint);
+    // Returns the average of all the data points seen so far.
+    float computeAverage() const;
+  };
+
+  // Map of Time-Series metrics tracked by the RPC Agent
+  std::unordered_map<std::string, std::unique_ptr<TimeSeriesMetricsTracker>>
```
Comment (lw) [inline torch/csrc/distributed/rpc/tensorpipe_agent.h:147]: What's the reason to use `unique_ptr` here? Wouldn't a dict containing `TimeSeriesMetricsTracker` object by value work?

**Pull request closed without merging**